hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/DirectInStoreCompactor.java

```diff
@@ -0,0 +1,90 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hbase.regionserver.compactions;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.io.compress.Compression;
+import org.apache.hadoop.hbase.io.hfile.CacheConfig;
+import org.apache.hadoop.hbase.io.hfile.HFileContext;
+import org.apache.hadoop.hbase.regionserver.HStore;
+import org.apache.hadoop.hbase.regionserver.HStoreFile;
+import org.apache.hadoop.hbase.regionserver.StoreFileWriter;
+import org.apache.yetus.audience.InterfaceAudience;
+
+@InterfaceAudience.Private
+public class DirectInStoreCompactor extends DefaultCompactor {
+  public DirectInStoreCompactor(Configuration conf, HStore store) {
+    super(conf, store);
+  }
+
+  @Override
+  protected StoreFileWriter initWriter(FileDetails fd, boolean shouldDropBehind, boolean major)
+    throws IOException {
+    // When all MVCC readpoints are 0, don't write them.
+    // See HBASE-8166, HBASE-12600, and HBASE-13389.
+    return createWriterInFamilyDir(fd.maxKeyCount,
+      major ? majorCompactionCompression : minorCompactionCompression,
+      fd.maxMVCCReadpoint > 0, fd.maxTagsLength > 0, shouldDropBehind);
+  }
+
+  private StoreFileWriter createWriterInFamilyDir(long maxKeyCount,
+      Compression.Algorithm compression, boolean includeMVCCReadpoint, boolean includesTag,
+        boolean shouldDropBehind) throws IOException {
+    final CacheConfig writerCacheConf;
+    // Don't cache data on write on compactions.
+    writerCacheConf = new CacheConfig(store.getCacheConfig());
+    writerCacheConf.setCacheDataOnWrite(false);
+
+    InetSocketAddress[] favoredNodes = null;
+    if (store.getHRegion().getRegionServerServices() != null) {
+      favoredNodes = store.getHRegion().getRegionServerServices().getFavoredNodesForRegion(
+        store.getHRegion().getRegionInfo().getEncodedName());
+    }
+    HFileContext hFileContext = store.createFileContext(compression, includeMVCCReadpoint,
+      includesTag, store.getCryptoContext());
+    Path familyDir = new Path(store.getRegionFileSystem().getRegionDir(),
+      store.getColumnFamilyDescriptor().getNameAsString());
+    StoreFileWriter.Builder builder = new StoreFileWriter.Builder(conf, writerCacheConf,
+      store.getFileSystem())
+      .withOutputDir(familyDir)
+      .withBloomType(store.getColumnFamilyDescriptor().getBloomFilterType())
+      .withMaxKeyCount(maxKeyCount)
+      .withFavoredNodes(favoredNodes)
+      .withFileContext(hFileContext)
+      .withShouldDropCacheBehind(shouldDropBehind)
+      .withCompactedFilesSupplier(() -> store.getCompactedFiles());
```

Comment:
Again, can be gotten from StoreContext. The purpose of the StoreContext was to avoid having to pass the HStore instance (and the ability to call HStore methods). Granted in this case you need to have access to the store object anyways to call createStoreFileAndReader so it's less important, but please keep in mind for future changes.

Reply:
Ack.